File: common/src/main/java/cc/turtl/chiselmon/ChiselmonStorage.java
package cc.turtl.chiselmon;

import cc.turtl.chiselmon.api.filter.FiltersUserData;
import cc.turtl.chiselmon.api.storage.ScopedData;
import cc.turtl.chiselmon.api.storage.StorageScope;
import cc.turtl.chiselmon.api.storage.adapter.GsonAdapter;
import cc.turtl.chiselmon.api.storage.adapter.H2Adapter;
import cc.turtl.chiselmon.feature.pc.PCUserData;
import cc.turtl.chiselmon.system.tracker.EncounterDatabase;
import cc.turtl.turtlshell.api.client.ClientEvents;
import kotlin.Unit;

import java.util.List;

/**
 * Central registry of all scoped data storage.
 */
public class ChiselmonStorage {
    public static final ScopedData<FiltersUserData> FILTERS = new ScopedData<>(
            GsonAdapter.of("filters.json", FiltersUserData.class, FiltersUserData::withDefaults)
    );
    public static final ScopedData<PCUserData> PC_SETTINGS = new ScopedData<>(
            GsonAdapter.of("pc.json", PCUserData.class, PCUserData::new)
    );
    public static final ScopedData<EncounterDatabase> ENCOUNTERS = new ScopedData<>(
            H2Adapter.of("encounters", EncounterDatabase::new, EncounterDatabase::flush, EncounterDatabase::close)
    );

    private static final List<ScopedData<?>> ALL = List.of(FILTERS, PC_SETTINGS, ENCOUNTERS);

    private static final int AUTOSAVE_INTERVAL_TICKS = 20 * 60 * 5;
    private static int tickCount = 0;

    /**
     * Registers auto saving + clear on disconnect.
     */
    public static void init() {

        // Save + clear world-scoped data on world leave
        ClientEvents.INSTANCE.getLEVEL_DISCONNECTED().subscribe(e -> {
            StorageScope world = StorageScope.currentWorld();
            if (world != null) {
                for (ScopedData<?> data : ALL) {
                    data.saveAndClear(world);
                }
            }
            return Unit.INSTANCE;
        });

        // Save everything on game close
        ClientEvents.INSTANCE.getGAME_STOPPING().subscribe(e -> {
            saveAll();
            return Unit.INSTANCE;
        });

        // Autosave every 5 minutes (only while in a world)
        ClientEvents.INSTANCE.getTICK_POST().subscribe(e -> {
            if (++tickCount >= AUTOSAVE_INTERVAL_TICKS) {
                tickCount = 0;
                saveAll();
            }
            return Unit.INSTANCE;
        });
    }

    /**
     * Saves all currently loaded data across all scopes.
     */
    public static void saveAll() {
        for (ScopedData<?> data : ALL) {
            data.saveAll();
        }
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/client/ChiselmonStorage.kt
package cc.turtl.chiselmon.client

import cc.turtl.chiselmon.api.filter.FiltersUserData
import cc.turtl.chiselmon.core.api.storage.Scope
import cc.turtl.chiselmon.core.api.storage.gsonData
import cc.turtl.chiselmon.core.api.storage.h2Data
import cc.turtl.chiselmon.client.feature.pc.PCUserData
import cc.turtl.chiselmon.system.tracker.EncounterDatabase
import cc.turtl.turtlshell.api.client.ClientEvents

/**
 * Central registry for scoped data storage.
 *
 *   val data = ChiselmonStorage.FILTERS[StorageScope.global()]
 */
object ChiselmonStorage {

    @JvmField
    val FILTERS = gsonData("filters.json", FiltersUserData::class.java) { FiltersUserData.withDefaults() }

    @JvmField
    val PC_SETTINGS = gsonData("pc.json", PCUserData::class.java, ::PCUserData)

    @JvmField
    val ENCOUNTERS = h2Data("encounters", ::EncounterDatabase, EncounterDatabase::flush, EncounterDatabase::close)

    private val all = listOf(FILTERS, PC_SETTINGS, ENCOUNTERS)

    private const val AUTOSAVE_INTERVAL_TICKS = 20 * 60 * 5
    private var tickCount = 0

    /** Registers autosave and world-leave cleanup. Call once during mod init. */
    fun init() {
        ClientEvents.LEVEL_DISCONNECTED.subscribe {
            Scope.currentWorld()?.let { world -> all.forEach { it.saveAndClear(world) } }
        }

        ClientEvents.GAME_STOPPING.subscribe {
            saveAll()
        }

        ClientEvents.TICK_POST.subscribe {
            if (++tickCount >= AUTOSAVE_INTERVAL_TICKS) {
                tickCount = 0
                saveAll()
            }
        }
    }

    /** Saves all currently loaded data across all scopes. */
    fun saveAll() = all.forEach { it.saveAll() }
}